Fix consexpr value declaration usage

arades79 opened the pull request:
Using `constexpr` to declare constants inside of non-constexpr function scope does not guarantee that the value will be inlined, and will use automatic storage (stack storage) unless the compiler optimizes it.

This is fixed by declaring constants as `constexpr static`, forcing the compiler to evaluate at compile time, and inlining the value keeping it off the stack. 

Different compilers and optimization values will behave differently. This is extremely unlikely to affect any release builds, but could reduce stack usage and *maybe* increase performance in debug builds, or builds using MSVC as it doesn't like to eagerly evaluate constant expressions.

[reference video for further explanation](https://www.youtube.com/watch?v=4pKtPWcl1Go)

I replaced all the relevant `constexpr` constant delcarations with `constexpr static`. All builds and tests pass. I tried a couple games and didn't see any immediate issues. Only tested on Linux.

Kelebek1 commented:
Can `constinit` be used here for the same effect? Not sure which is preferable, but I'd prefer to read constinit than constexpr static personally.

arades79 commented:
Unfortunately no. I hadn't actually used constinit before, and it's not completely clear [in the docs](https://en.cppreference.com/w/cpp/language/constinit). However, it appears constinit 1. doesn't actually create constant variables, it just asserts that the initializer is a constant expression; 2. also requires static or thread_local.

[example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:18,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:5,endLineNumber:11,positionColumn:5,positionLineNumber:11,selectionStartColumn:5,selectionStartLineNumber:11,startColumn:5,startLineNumber:11),source:'//+Type+your+code+here,+or+load+an+example.%0Aconstexpr+int+square(int+num)+%7B%0A++++constexpr+int+i+%3D+1%3B%0A++++return+num+*+num%3B%0A%7D%0A%0Aint+main()+%7B%0A++++%0A++++//+exists+on+the+stack%0A++++constexpr+auto+a1+%3D+square(1)%3B%0A++++%0A++++//+totally+inlined%0A++++static+constexpr+auto+a2+%3D+square(2)%3B%0A++++%0A++++//+not+constant%0A++++static+constinit+auto+a3+%3D+square(4)%3B%0A%0A++++a3+%2B%3D+5%3B%0A%0A++++return++a1%2Ba2%2Ba3%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:26.771488728337253,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g122,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'0'),flagsViewOpen:'1',fontScale:18,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O0+-std%3Dc%2B%2B20',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+12.2+(Editor+%231)',t:'0')),k:39.89517793832942,l:'4',m:100.00000000000003,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+12.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+12.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Kelebek1 commented:
@lioncash thoughts?

SciresM commented:
For the kernel, at least, the changed declarations aren't constexpr static. In official kernel, these constants are sometimes stack variables depending on optimization flags/system version, and they never go in .rodata (which is what static would do when not inlined in move constants).

Just noting that this is a divergence, and I advise against changing those cases at least.

arades79 commented:
I would think that could only cause a difference if there were applications that did some direct stack manipulation to probe values which is absurdly unsound. That being said you certainly know Horizon kernel behavior better than I :P.

In that case, for clarity those specific variables probably shouldn't be marked constexpr, and just regular const instead - with a comment saying that they're marked to match official behavior. 

I feel that when you see a constexpr declaration it's surprising that it causes a stack allocation, and forces a load for what should be an immediate value.

liamwhite commented:
As a matter of purely style preference I try to avoid constexpr static (except for static class members, where it is otherwise required). Unless you were overflowing the stack it's hard to imagine this would cause a noticeable difference in performance.

lioncash reviewed (changes requested):
Constants of primitive type will almost always be inlined and there's no real reason to make these static. Even barebones optimizations with -O1 will fold primitive constants. The only case where `static` would be beneficial is with `std::array` instances if they're large, since I think MSVC will construct those on the stack, so a large section of these changes are unnecessary.



Our convention for those cases is `static constexpr` rather than `constexpr static` since it puts the lifetime-related keyword/the important bit first.

I don't think marking every single constant as static when it's not shown the specific declarations in particular are even problematic to begin with.

arades79 commented:
Okay!

I removed the statics for integral types where they'll definitely be inlined. I kept the statics for object types, especially larger-than-pointer sized objects. 

There were a couple instances in the vulkan files where the stack allocated constexpr objects were being passed by pointer to constructors. I didn't look too deep in the implementation to see how these pointers get handled, but to be safe I kept those as static as well to avoid the possible UB of a dangling reference.

I also changed one of the Gdb functions to use string_view instead of string since it's only ever passed as a string_view anyway. Before it was allocating for the huge xml string even though it is constant.

Then I rebased to pickup the new fixes.

liamwhite commented:
Seems fine except for the change to the gdb server - if you declare a method as constexpr, its body must be visible at the point of use.

lioncash reviewed (approved):
LGTM
